src/libraries/System.Text.Json/Common/JsonSimpleNamingPolicy.cs

```diff
@@ -0,0 +1,147 @@
+
+                if (_lowercase)
+                {
+                    word.ToLowerInvariant(destination);
+                }
+                else
+                {
+                    word.ToUpperInvariant(destination);
+                }
+
+                resultLength += word.Length;
+            }
+
+            int first = 0;
+            ReadOnlySpan<char> chars = name.AsSpan();
+            CharCategory previousCategory = CharCategory.Boundary;
+
+            for (int index = 0; index < chars.Length; index++)
+            {
+                char current = chars[index];
+                UnicodeCategory currentCategoryUnicode = char.GetUnicodeCategory(current);
+
+                if (currentCategoryUnicode == UnicodeCategory.SpaceSeparator ||
+                    currentCategoryUnicode >= UnicodeCategory.ConnectorPunctuation &&
+                    currentCategoryUnicode <= UnicodeCategory.OtherPunctuation)
+                {
+                    WriteWord(ref result, chars.Slice(first, index - first));
+
+                    previousCategory = CharCategory.Boundary;
+                    first = index + 1;
+
+                    continue;
+                }
+
+                if (index + 1 < chars.Length)
+                {
+                    char next = chars[index + 1];
+                    CharCategory currentCategory = currentCategoryUnicode switch
+                    {
+                        UnicodeCategory.LowercaseLetter => CharCategory.Lowercase,
+                        UnicodeCategory.UppercaseLetter => CharCategory.Uppercase,
+                        _ => previousCategory
+                    };
+
+                    if (currentCategory == CharCategory.Lowercase && char.IsUpper(next) ||
+                        next == '_')
+                    {
+                        WriteWord(ref result, chars.Slice(first, index - first + 1));
+
+                        previousCategory = CharCategory.Boundary;
+                        first = index + 1;
+
+                        continue;
+                    }
+
+                    if (previousCategory == CharCategory.Uppercase &&
+                        currentCategoryUnicode == UnicodeCategory.UppercaseLetter &&
+                        char.IsLower(next))
+                    {
+                        WriteWord(ref result, chars.Slice(first, index - first));
+
+                        previousCategory = CharCategory.Boundary;
+                        first = index;
+
+                        continue;
+                    }
+
+                    previousCategory = currentCategory;
+                }
+            }
+
+            WriteWord(ref result, chars.Slice(first));
+
+            name = result
+                .Slice(0, resultLength)
+                .ToString();
+
+            if (buffer is not null)
+                ArrayPool<char>.Shared.Return(buffer);
+
+            return name;
+        }
+
+        private enum CharCategory
```

Comment:
Maybe we could use a byte instead of (by default int) here (unless we have disadvantages like performance loss)

```suggestion
        private enum CharCategory : byte
```

Reply:
I don't think it's super important in this context.